pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,587 @@
+                Error::<T>::CannotCancelInProcessAuction
+            );
+
+            // List nft as not for sale
+            T::NFTHandler::set_listed_for_sale(nft_id, false)?;
+
+            // Remove auction from storage
+            Auctions::<T>::remove(nft_id);
+
+            // Emit auction canceled event
+            Self::deposit_event(Event::AuctionCancelled {
+                nft_id,
+                marketplace_id: auction.marketplace_id,
+                creator: who,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        #[transactional]
+        pub fn add_bid(
+            origin: OriginFor<T>,
+            nft_id: NFTId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            // fetch the data of given nftId
+            let nft_data = T::NFTHandler::get_nft(nft_id).ok_or(Error::<T>::NFTIdInvalid)?;
+
+            // fetch data of auction that lists NFT
+            let auction = Auctions::<T>::get(nft_id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+            // ensure the nft is in listed for sale state
+            ensure!(
+                nft_data.listed_for_sale == true,
+                Error::<T>::NFTNotListedForSale
+            );
+
+            // ensure the caller is not the owner of NFT
+            ensure!(
+                nft_data.owner != who.clone(),
+                Error::<T>::OwnerCannotCreateBid
+            );
+
+            // ensure the auction period has commenced
+            ensure!(
+                auction.start_block < current_block,
+                Error::<T>::AuctionNotStarted
+            );
+
+            // ensure the auction period has not ended
+            ensure!(auction.end_block >= current_block, Error::<T>::AuctionEnded);
+
+            // ensure the bid amount is greater than start price
+            ensure!(auction.start_price < amount, Error::<T>::InvalidBidAmount);
+
+            // ensure the bid is larger than the current highest bid
+            if let Some(current_highest_bid) = auction.bidders.get_highest_bid() {
+                ensure!(amount > current_highest_bid.1, Error::<T>::InvalidBidAmount);
+            }
+
+            // reject if user already has a bid
+            ensure!(
+                auction.bidders.find_bid(&who).is_none(),
+                Error::<T>::UserBidAlreadyExists
+            );
+
+            // transfer funds from caller
+            T::Currency::transfer(&who, &Self::account_id(), amount, KeepAlive)?;
+
+            // add bid to storage
+            Auctions::<T>::try_mutate(nft_id, |maybe_auction| -> DispatchResult {
+                // should not panic when unwrap since already checked above
+                let auction = maybe_auction.as_mut().unwrap();
```

Comment:
To continue the discussion...
The only place where unwrap is allowed is in tests or benchmarks. Here you need to return an error if the value is not present.

Reply:
It doesn't matter if you have already check it once or twice. Unwraps and panics are forbidden in non-test/non-benchmark code

Reply:
fixed